### PR TITLE
OPA: improve error messages returned to the end user

### PR DESCRIPTION
**Pull request by aliculPix4D**
Draft PR just to fish for some initial feedback and to show what changes are required here and in our ci-for-concourse repo..

Related to: https://github.com/Pix4D/ci-for-concourse/pull/18

### CODE REVIEW

File: atc/api/policychecker/policychecker_suite_test.go [hunk 1]
```diff
@@ -10,7 +10,7 @@ import (
 	"testing"
 )
 
-func TestAccessor(t *testing.T) {
```
Comment (aliculPix4D) [inline atc/api/policychecker/policychecker_suite_test.go:13]: seems as copy pasta...

File: atc/policy/opa/agent_suite_test.go [hunk 2]
```diff
@@ -7,7 +7,7 @@ import (
 	. "github.com/onsi/gomega"
 )
 
-func TestVault(t *testing.T) {
```
Comment (aliculPix4D) [inline atc/policy/opa/agent_suite_test.go:10]: seems as copy pasta...

File: atc/policy/opa/opa.go [hunk 3]
```diff
@@ -60,23 +60,23 @@ func (c opa) Check(input policy.PolicyCheckInput) (policy.PolicyCheckResult, err
...
 
 	body, err := io.ReadAll(resp.Body)
 	if err != nil {
-		return nil, fmt.Errorf("opa returned no response: %s", err.Error())
+		return nil, fmt.Errorf("opa server returned no response: %s", err.Error())
 	}
 
 	result, err := ParseOpaResult(body, c.config)
 	if err != nil {
-		return nil, fmt.Errorf("parsing opa results: %w", err)
```
Comment (aliculPix4D) [inline atc/policy/opa/opa.go:79]: we only parse OPA results if the response is a valid JSON...

**Comment by aliculPix4D** [comment]
@marco-m-pix4d fishing for feedback

**Comment by aliculPix4D** [comment]
Two things I forgot to mention:
1. in theory the same confusing error can also arise from (both for json and yaml):
https://github.com/Pix4D/concourse/blob/b4180829d581cb774f21260c138f502e61c8c641/atc/api/policychecker/checker.go#L65
https://github.com/Pix4D/concourse/blob/b4180829d581cb774f21260c138f502e61c8c641/atc/api/policychecker/checker_test.go#L126
I choose not to change anything there because in my understanding this can happen only if the request is crafted wrongly by one of the Concourse internal components or fly... which would mean there is an existing issue in the code.
Ok, sure it can also happen if user uses curl to craft request and targets the internal API request directly (without going through fly). I mean we do that in reaper and and skyguide but not e.g. to set the pipeline etc..

In any case, its easy to do the same change here so, opinions?

2. the PR as it is may or may not be enough... the end goal is to improve the error msg to the end user... The question is: is this error reported back to the use if e.g. is using fly? I will test this part now... I already know for example that `fly expose-pipeline` does not report any meaningful error back to the user, just prints `forbidden` with no reason (on the othehand you see the response in the reason in the WebUI)... This was one of the problems when I started with OPA actually because I was testing first with ExposePipeline rule... For other commands, like fly set-pipeline I need to ensure that this msg reaches the end user in full... I don't remember seeing anything else than `invalid character ....` but looking already at existing code `parsing opa results: invalid character`... Anyway, I just wanted to make it clear that there is more to check and test here...

File: atc/api/policychecker/policychecker_suite_test.go (hunk 1 above)
Comment (marco-m-pix4d) [reply atc/api/policychecker/policychecker_suite_test.go:13]: Ah yes. We contributing upstream, we can even split this PR in 2. The first one with this fix and the equivalent you did below. At least such PR should be accepted without problems.

File: atc/policy/opa/opa.go [hunk 4]
```diff
@@ -60,23 +60,23 @@ func (c opa) Check(input policy.PolicyCheckInput) (policy.PolicyCheckResult, err
 	client.Timeout = c.config.Timeout
 	resp, err := client.Do(req)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("opa server connection error: %s", err)
```
Comment (marco-m-pix4d) [inline atc/policy/opa/opa.go:63]: This is a message, names of things should be correct.
The name of the service is OPA (uppercase).
We don't want to use the word "error" in error messages.
I would simply say `fmt.Errorf("OPA server: connecting: %s", err)`

File: atc/policy/opa/opa.go [hunk 5]
```diff
@@ -60,23 +60,23 @@ func (c opa) Check(input policy.PolicyCheckInput) (policy.PolicyCheckResult, err
...
 	}
 	defer resp.Body.Close()
 
 	statusCode := resp.StatusCode
 	if statusCode != http.StatusOK {
-		return nil, fmt.Errorf("opa returned status: %d", statusCode)
+		return nil, fmt.Errorf("opa server returned status: %d", statusCode)
```
Comment (marco-m-pix4d) [inline atc/policy/opa/opa.go:69]: OPA

File: atc/policy/opa/opa.go [hunk 6]
```diff
@@ -60,23 +60,23 @@ func (c opa) Check(input policy.PolicyCheckInput) (policy.PolicyCheckResult, err
...
 	}
 
 	body, err := io.ReadAll(resp.Body)
 	if err != nil {
-		return nil, fmt.Errorf("opa returned no response: %s", err.Error())
+		return nil, fmt.Errorf("opa server returned no response: %s", err.Error())
```
Comment (marco-m-pix4d) [inline atc/policy/opa/opa.go:74]: OPA

Comment (marco-m-pix4d) [reply atc/policy/opa/opa.go:74]: Ah. The message is wrong: it could have returned part of the response.
No need for err.Error(): the fmt functions know about errors.

`fmt.Errorf("OPA server: reading response body: %s", err)`

File: atc/policy/opa/opa.go (hunk 3 above)
Comment (marco-m-pix4d) [reply atc/policy/opa/opa.go:79]: opa -> OPA

Comment (marco-m-pix4d) [reply atc/policy/opa/opa.go:79]: I understand your reasoning, but I realized the following.
The function name is `ParseOpaResult`. If we start using the name "response", we just add confusion.
See also the parameter names: `ParseOpaResult(bytesResult []byte, ...`
Better to leave as-is (and use OPA uppercase everywhere)

File: atc/policy/opa/result.go [hunk 7]
```diff
@@ -30,7 +30,7 @@ func ParseOpaResult(bytesResult []byte, opaConfig OpaConfig) (opaResult, error)
 	var results vars.StaticVariables
 	err := json.Unmarshal(bytesResult, &results)
 	if err != nil {
-		return opaResult{}, err
+		return opaResult{}, fmt.Errorf("JSON parsing: error: %s. Hint: response from opa server is not a valid JSON", err)
```
Comment (marco-m-pix4d) [inline atc/policy/opa/result.go:33]: Careful. This is changing the API. Before it was returning the specific error type. Now it is returning a generic error type. Instead, use the `%w` verb to preserve the type.

File: atc/policy/opa/opa_test.go [hunk 8]
```diff
@@ -205,7 +205,7 @@ var _ = Describe("OPA Policy Checker", func() {
 		It("should return error", func() {
 			result, err := agent.Check(policy.PolicyCheckInput{})
 			Expect(err).To(HaveOccurred())
-			Expect(err.Error()).To(ContainSubstring("invalid character 'h' looking for beginning of value"))
```
Comment (marco-m-pix4d) [inline atc/policy/opa/opa_test.go:208]: This is so ironic I thought you invented it.

File: atc/policy/opa/result.go (hunk 7 above)
Comment (marco-m-pix4d) [reply atc/policy/opa/result.go:33]: We can simplify the message and still be very clear

```
fmt.Errorf("parsing JSON: %w", err)
```

Since the message would become
```
parsing JSON: invalid character 'h' looking for beginning of value
```
it is clear that it is a JSON parsing error, thus by definition not valid JSON.

**Review by marco-m-pix4d — DISMISSED** [review]
Looks good. It was easier than I though, better like this.

**Comment by aliculPix4D** [comment]
> Looks good. It was easier than I though, better like this.

@marco-m-pix4d please read also: https://github.com/Pix4D/concourse/pull/796#issuecomment-2373736504

**Comment by marco-m-pix4d** [comment]
> > Looks good. It was easier than I though, better like this.
> 
> @marco-m-pix4d please read also: [#796 (comment)](https://github.com/Pix4D/concourse/pull/796#issuecomment-2373736504)

Ah OK. The root cause is that the Go json library sometimes doesn't prefix its error messages with "json", and sometimes it does, so we even risk stuttering ("parsing JSON: json: xxx") :-(

So we can do the following:

Grep for all "invalid character 'h' looking for beginning of value`" in the tests for this area of the code and see if we can prefix the error with "parsing JSON" as we are doing here.

**Comment by aliculPix4D** [comment]
Update: I still didn't manage to fully fix this...

I don't know why I thought set-pipeline command was working...
Screenshot from 2024-09-26 09-26-27

it certainly does not work.  If I add the --verbose flag I see the full error msg... 

So yes, to reach the end goal (improve the error msg to the end user) I must sort this with fly also...

**Comment by aliculPix4D** [comment]
So I managed to solve the issue with both (maybe all commands)... As I expected the difference of behaviour between commands arrived from usage of different clients:
- some commands like set-pipeline use new httpClient
https://github.com/Pix4D/concourse/blob/b4180829d581cb774f21260c138f502e61c8c641/go-concourse/concourse/configs.go#L68
(FYI its even more crazy than that: for fetching pipeline config (GET) you use deprecated connection client while for creating/updating (PUT) of the config an HTTP client is used)

- other commands still use deprecated connection client
https://github.com/Pix4D/concourse/blob/b4180829d581cb774f21260c138f502e61c8c641/go-concourse/concourse/pipelines.go#L237

1. for the connection client the cause and solution is quite straightforward and exactly what we discussed during standup:

currently if the response is forbidden by the e.g. OPA just the status code is returned without any reason.
```
  if response.StatusCode == http.StatusForbidden {
    return ErrForbidden
  }
```
what we actually need is:
```
  if response.StatusCode == http.StatusForbidden {
    body, _ := io.ReadAll(response.Body)
    return ForbiddenError{
      Reason: string(body),
    }
  }
```
to print also the reason from the response body.

Note that this simple fix actually affect many many commands. I would even say majority of the commands suffered from the same issue.

2. for the set pipeline command things are a bit different than we expected

Even though we fixed the server side already, I was still saying: `invalid character 'h' looking for beginning of value`. OK, it was clear its another JSON parser, but which one?

The cause of the error arrives from:
https://github.com/Pix4D/concourse/blob/b4180829d581cb774f21260c138f502e61c8c641/go-concourse/concourse/configs.go#L97

So my explanation is this: originally you enter the `case http.StatusBadRequest:` when you fail to set/update the pipeline due to some configuration errors in the pipeline itself, so the Concourse server rejects it and returns a StatusBadRequest.
But they wanted to give a proper error msg to the end user so they decided to extract it from the body and convert it to JSON so that they can do something like:
```
var validationErr atc.SaveConfigResponse
err = json.Unmarshal(body, &validationErr)
return InvalidConfigError{Errors: validationErr.Errors}
```
The problem is that now with OPA (but maybe also due to other cases) one can not expect that body is in JSON format, and especially you should not expect that it can be converted to atc.SaveConfigResponse type.

so when server was returning a body with an error msg like:
```
Body: policy check error: parsing opa response: JSON parsing: error: invalid character 'h' looking for beginning of value. Hint: response from opa server is not a valid JSON
```

we were just returning the error msg from the JSON parser, which is this time:
> invalid character 'p' looking for beginning of value

even though OPA policy checker returned:
`policy check error: parsing opa response: JSON parsing: error: invalid character 'h' looking for beginning of value`
because OPA server return a body `hello` we have: `invalid character 'h' looking for beginning of value`, but because the second JSON parser started to parse a body:
> policy check error: parsing opa response: JSON parsing: error: invalid character 'h' looking for beginning of value. Hint: response from opa server is not a valid JSON

we have:
> invalid character 'p' looking for beginning of value

seen by the end user...

My first attempt to fix this is:
```
    var validationErr atc.SaveConfigResponse
    err = json.Unmarshal(body, &validationErr)
    if err != nil {
      return false, false, []ConfigWarning{}, internal.UnexpectedResponseError{
        StatusCode: response.StatusCode,
        Status:     response.Status,
        Body:       string(body),
      }
    }
```
This prints what we want to the end user, but we lost the actual error msg from the json.Unmarshal just above...
Personally, I am fine with that... I am just not sure yet this is the best solution.

File: atc/policy/opa/agent_suite_test.go (hunk 2 above)
Comment (aliculPix4D) [reply atc/policy/opa/agent_suite_test.go:10]: Fixed...

File: atc/policy/opa/opa.go (hunk 3 above)
Comment (aliculPix4D) [reply atc/policy/opa/opa.go:79]: Fixed...

File: atc/policy/opa/opa_test.go (hunk 8 above)
Comment (aliculPix4D) [reply atc/policy/opa/opa_test.go:208]: Now I know it's even worse... The code from current master would craft this error msg:
> parsing opa results: invalid character 'h' looking for beginning of value

and this is not the error msg end-user was getting... but instead end user saw only:
> invalid character 'p' looking for beginning of value

because another JSON parser got the body starting with `parsing...`
:)

Unbelievable...

**Comment by aliculPix4D** [comment]
Summary of changes:

fly set-pipeline when OPA responded with 200 but with non-json body

What user saw before:

```
$ fly -t main set-pipeline -c pipeline.yml -p test

jobs:
  job job-1 has been added:
+ name: job-1
+ plan: null
  
pipeline name: test

apply configuration? [yN]: y

error: invalid character 'p' looking for beginning of value
```

What it sees now?

```
$ fly -t main set-pipeline -c pipeline.yml -p test

jobs:
  job job-1 has been added:
+ name: job-1
+ plan: null
  
pipeline name: test

apply configuration? [yN]: y

error: Unexpected Response
Status: 400 Bad Request
Body: policy check error: parsing OPA results: parsing JSON: invalid character 'h' looking for beginning of value
```


fly expose-pipeline (and many other command) when OPA returned 200 and denied the action (correct json body).


What user saw before:

```
$ fly -t main expose-pipeline -p test
error: forbidden
```

What it sees now?

```
$ fly -t main expose-pipeline -p test
error: forbidden: policy check failed: 
 * because we said so
```

File: atc/policy/opa/result.go [hunk 9]
```diff
@@ -30,7 +30,7 @@ func ParseOpaResult(bytesResult []byte, opaConfig OpaConfig) (opaResult, error)
 	var results vars.StaticVariables
 	err := json.Unmarshal(bytesResult, &results)
 	if err != nil {
-		return opaResult{}, err
+		return opaResult{}, fmt.Errorf("parsing JSON: %w", err)
```
Comment (aliculPix4D) [inline atc/policy/opa/result.go:33]: Personally, I am still not 100% happy with the error msg user sees:
> policy check error: parsing OPA results: parsing JSON: invalid character 'h' looking for beginning of value

for sure its better than before:
> invalid character 'p' looking for beginning of value

but I am somehow lacking the information that this is because the response body from the OPA server is not a valid JSON..  or in other words that the problem is because OPA server being misconfigured and sending something unexpected...

File: atc/policy/opa/result.go (hunk 7 above)
Comment (aliculPix4D) [reply atc/policy/opa/result.go]: Done...

File: atc/policy/opa/opa.go (hunk 5 above)
Comment (aliculPix4D) [reply atc/policy/opa/opa.go]: Done..

File: atc/policy/opa/opa.go (hunk 4 above)
Comment (aliculPix4D) [reply atc/policy/opa/opa.go]: Done as suggested.

File: atc/policy/opa/opa.go [hunk 10]
```diff
@@ -60,23 +60,23 @@ func (c opa) Check(input policy.PolicyCheckInput) (policy.PolicyCheckResult, err
...
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("OPA server: connecting: %w", err)
 	}
 	defer resp.Body.Close()
 
 	statusCode := resp.StatusCode
 	if statusCode != http.StatusOK {
-		return nil, fmt.Errorf("opa returned status: %d", statusCode)
+		return nil, fmt.Errorf("OPA server returned status: %d", statusCode)
```
Comment (aliculPix4D) [inline atc/policy/opa/opa.go:69]: do we want to do something with the response body here?

**Comment by aliculPix4D** [comment]
@Pix4D/integration ready for review...

I split the work into 4 commits. The two very important ones (third and fourth) have longer commit msg explaining the reasoning why they are required (shorter version of my PR comments)

File: atc/api/policychecker/policychecker_suite_test.go (hunk 1 above)
Comment (aliculPix4D) [reply atc/api/policychecker/policychecker_suite_test.go:13]: My plan:

- merge it to our fork as a single PR to our master (and release branch)... I will now read our instructions on release procedure
- after we release and deploy I plan to prepare the PRs to upstream
- then cherry pick specific commits and open 2 PRs upstream if your think its necessary to split this changes

File: atc/policy/opa/opa_test.go (hunk 8 above)
Comment (marco-m-pix4d) [reply atc/policy/opa/opa_test.go:208]: > because another JSON parser got the body starting with parsing...

Ah. Can we fix also this properly ?

File: go-concourse/concourse/internal/connection_test.go [hunk 11]
```diff
@@ -353,7 +353,7 @@ var _ = Describe("ATC Connection", func() {
 				})
 			})
 
-			Describe("403 response", func() {
+			Describe("HTTPAgent: 403 response", func() {
```
Comment (aliculPix4D) [inline go-concourse/concourse/internal/connection_test.go:356]: they had a test  for 403 response for what they call `HTTPagent` client, but not for what they call `Connection` client

File: atc/policy/opa/opa_test.go (hunk 8 above)
Comment (aliculPix4D) [reply atc/policy/opa/opa_test.go:208]: We did... 4th commit...

**Comment by aliculPix4D** [comment]
@Pix4D/integration one more note... I think reading carefully:
https://github.com/Pix4D/concourse/pull/796#issuecomment-2376498057
https://github.com/Pix4D/concourse/pull/796#issuecomment-2377000929

before reviewing will help you a lot to understand what is going on...

File: atc/policy/opa/result.go (hunk 9 above)
Comment (marco-m-pix4d) [reply atc/policy/opa/result.go:33]: If you think about it, from the point of view of the end user (the dev setting the pipeline), nothing changes:

- `policy check error: parsing OPA results: parsing JSON: invalid character 'h' looking for beginning of value`
- `policy check error: parsing OPA results: parsing JSON: invalid character 'h' looking for beginning of value. Hint: targeted OPA server seems to be misconfigured`

In both cases, the user will ask for help on the channel: he cannot do anything about it and he is still confused.

I really resist this "Hint" because it is adding a patch to something that is broken already.

If we really want to give a better message to the user, I think we can, because we control the system (Concourse). This is different from cogito: the system is GitHub that we do not control.

So we _should_ be able to rework some more logic how Concourse interacts with the OPA server and detect the real problem. Once we do that, then we could report something like

- `policy check error: OPA server misconfigured`

Now, I don't even understand what we mean by "misconfigured".:
- Do we mean that the OPA server configuration is wrong? (so: independently from Concourse).
- Do we mean that the Concourse settings about the OPA server is wrong?

To summarize, I think that this PR already made the situation way better and we should merge and deploy.

I am OK if we want to create a ticket, with some of the discussion here, to go more at the root of the problem (assuming I am not misunderstanding something).

Comment (marco-m-pix4d) [reply atc/policy/opa/result.go:33]: Said in another way: this PR is too big to follow in details, at least for me. All the discussions here are useful, but we can still do the secondary improvements in a separate ticket, I think.

File: atc/policy/opa/opa.go (hunk 10 above)
Comment (marco-m-pix4d) [reply atc/policy/opa/opa.go:69]: Fine if we do as in other places and add also the body to the error.

**Review by marco-m-pix4d — APPROVED** [review]
I left my detailed reasoning in my latest inline comment.
I think we can proceed (unless I am missing something).
Thanks!

File: go-concourse/concourse/configs.go [hunk 12]
```diff
@@ -96,7 +96,11 @@ func (team *team) CreateOrUpdatePipelineConfig(pipelineRef atc.PipelineRef, conf
 		var validationErr atc.SaveConfigResponse
 		err = json.Unmarshal(body, &validationErr)
 		if err != nil {
-			return false, false, []ConfigWarning{}, err
+			return false, false, []ConfigWarning{}, internal.UnexpectedResponseError{
```
Comment (aliculPix4D) [inline go-concourse/concourse/configs.go:99]: @marco-m-pix4d I am curious to hear/read your opinion on this part...

if you follow closely: 
- before we were returning the `err` which is a useless error from the JSON parse or L97...
- now we return `internal.UnexpectedResponseError` which I craft using response body and status code, but I never actually use the useless `err` anywhere... I think it is not needed to append this `err.Error()` because this error msg would always look the same `invalid character`.

Personally, I was uncomfortable just ignoring the `err` like that... We usually don't have this practice... but I don't think it would bring any value appending it.

What do you think?

Comment (marco-m-pix4d) [reply go-concourse/concourse/configs.go:99]: I see the reasoning. As usual, the answer is that it depends.

The problem is that json.Unmarshal can, at least in theory, return any type of error, not only parse errors.
If the json package had an error sentinel, say ErrParse, then one could say:
I check for json.ErrParse. If it matches, then I do not pass it to internal.UnexpectedResponseError{}.
If on the other hand it doesn't match, then it can be helpful to know what has happened, so in this case I pass it to  internal.UnexpectedResponseError{}.

Also if json.ErrParse existed, doing the reasoning just above would still be risky, because it is possible that a new version of the `json` package added another error sentinel, and so on...

So my approach is to almost always include errors. If I know like in this case that it can be confusing, I add it in parenthesis "(%s)" or similar...

**Comment by aliculPix4D** [comment]
As discussed in video chat, I am closing this. I will split it in smaller separate PRs as suggested.
